// @screencloud/signage-sdk/build/mixed/MixedPlayer.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var app_core_1 = require("@screencloud/app-core");
var __1 = require("..");
/**
 * @experimental
 *
 * non-constructor methods are copied over from EditorPlayer (methods that exists in AppPlayer already are omitted).
 */
var MixedPlayer = /** @class */ (function (_super) {
    __extends(MixedPlayer, _super);
    function MixedPlayer(handlers, bridge) {
        return _super.call(this, handlers || {}, bridge || new app_core_1.PostMessageBridge()) || this;
    }
    MixedPlayer.prototype.emitConfigUpdateAvailable = function () {
        this.emit({
            type: __1.EditorPlayerMessages.ConfigUpdateAvailable.type,
        });
    };
    MixedPlayer.prototype.onConfigUpdated = function (handler) {
        return this.on(__1.EditorMessages.ConfigUpdated.type, handler);
    };
    MixedPlayer.prototype.onRequestConfigUpdate = function (handler) {
        return this.on(__1.EditorMessages.RequestConfigUpdate.type, handler);
    };
    MixedPlayer.prototype.requestFileUpload = function () {
        return this.request({
            type: __1.EditorPlayerMessages.RequestFileUpload.type,
        });
    };
    MixedPlayer.prototype.requestAuthToken = function (payload) {
        return this.request({
            payload: payload,
            type: __1.EditorPlayerMessages.RequestAuthToken.type,
        });
    };
    MixedPlayer.prototype.requestMedia = function (payload) {
        return this.request({
            payload: payload,
            type: __1.EditorPlayerMessages.RequestMedia.type,
        });
    };
    return MixedPlayer;
}(__1.AppPlayer));
exports.MixedPlayer = MixedPlayer;
