// @screencloud/signage-sdk/build/editor/EditorPlayer.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var app_core_1 = require("@screencloud/app-core");
var messages_1 = require("../messages");
var EditorPlayer = /** @class */ (function (_super) {
    __extends(EditorPlayer, _super);
    function EditorPlayer(handlers, bridge) {
        return _super.call(this, handlers || {}, bridge || new app_core_1.PostMessageBridge()) || this;
    }
    EditorPlayer.prototype.emitInitialized = function () {
        this.emit({
            type: messages_1.EditorPlayerMessages.Initialized.type,
        });
    };
    EditorPlayer.prototype.emitConfigUpdateAvailable = function () {
        this.emit({
            type: messages_1.EditorPlayerMessages.ConfigUpdateAvailable.type,
        });
    };
    EditorPlayer.prototype.onInitialize = function (handler) {
        return this.on(messages_1.EditorMessages.Initialize.type, handler);
    };
    EditorPlayer.prototype.onConfigUpdated = function (handler) {
        return this.on(messages_1.EditorMessages.ConfigUpdated.type, handler);
    };
    EditorPlayer.prototype.onRequestConfigUpdate = function (handler) {
        return this.on(messages_1.EditorMessages.RequestConfigUpdate.type, handler);
    };
    EditorPlayer.prototype.requestFileUpload = function () {
        return this.request({
            type: messages_1.EditorPlayerMessages.RequestFileUpload.type,
        }, {
            timeout: -1,
        });
    };
    EditorPlayer.prototype.requestAuthToken = function (payload) {
        return this.request({
            payload: payload,
            type: messages_1.EditorPlayerMessages.RequestAuthToken.type,
        });
    };
    EditorPlayer.prototype.requestMedia = function (payload) {
        return this.request({
            payload: payload,
            type: messages_1.EditorPlayerMessages.RequestMedia.type,
        }, {
            timeout: -1,
        });
    };
    return EditorPlayer;
}(app_core_1.MessageApp));
exports.EditorPlayer = EditorPlayer;
