// @screencloud/signage-sdk/build/editor/EditorPlayerHost.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var app_core_1 = require("@screencloud/app-core");
var messages_1 = require("../messages");
var EditorPlayerHost = /** @class */ (function (_super) {
    __extends(EditorPlayerHost, _super);
    function EditorPlayerHost(handlers, bridge) {
        return _super.call(this, handlers || {}, bridge || new app_core_1.PostMessageBridge()) || this;
    }
    EditorPlayerHost.prototype.connect = function () {
        return _super.prototype.connect.call(this, true);
    };
    EditorPlayerHost.prototype.emitInitialize = function (payload) {
        this.emit({
            payload: payload,
            type: messages_1.EditorMessages.Initialize.type,
        });
    };
    EditorPlayerHost.prototype.emitConfigUpdated = function () {
        this.emit({
            type: messages_1.EditorMessages.ConfigUpdated.type,
        });
    };
    EditorPlayerHost.prototype.onInitialized = function (handler) {
        return this.on(messages_1.EditorPlayerMessages.Initialized.type, handler);
    };
    EditorPlayerHost.prototype.onConfigUpdateAvailable = function (handler) {
        return this.on(messages_1.EditorPlayerMessages.ConfigUpdateAvailable.type, handler);
    };
    EditorPlayerHost.prototype.onRequestMedia = function (handler) {
        return this.on(messages_1.EditorPlayerMessages.RequestMedia.type, handler);
    };
    EditorPlayerHost.prototype.onRequestAuthToken = function (handler) {
        return this.on(messages_1.EditorPlayerMessages.RequestAuthToken.type, handler);
    };
    EditorPlayerHost.prototype.onRequestFileUpload = function (handler) {
        return this.on(messages_1.EditorPlayerMessages.RequestFileUpload.type, handler);
    };
    EditorPlayerHost.prototype.requestConfigUpdate = function (payload) {
        return this.request({
            payload: payload,
            type: messages_1.EditorMessages.RequestConfigUpdate.type,
        });
    };
    return EditorPlayerHost;
}(app_core_1.MessageApp));
exports.EditorPlayerHost = EditorPlayerHost;
